### DISCUSSION TITLE
Cannot dereference pointer to Alpha_shape_3

### DISCUSSION THREAD

**Discussion by stla** [CGAL/cgal#6989] [Q&A]
Hello,

I have a poor knowledge in C/C++ and I don't know whether I'm doing something bad.

I have a R pointer (`Rcpp:Xptr`) to an `Alpha_shape_3`. I can get the ordinary pointer to the `Alpha_shape_3` from it, but when I try to dereference it, the code does not compile : it throws an error *"is private within this context"*.

```cpp
void optimalAlpha_cpp(Rcpp::XPtr<Alpha_shape_3> as_xptr, int nc) {
  Alpha_shape_3* as_ptr = as_xptr.get();
  Alpha_shape_3 as = *as_ptr; // error: is private within this context
  Alpha_iterator opt = as.find_optimal_alpha(nc);
  Rcpp::Rcout << *opt;
}
```

The following code compiles but it returns a wrong optimal alpha which makes no sense (a huge number):

```cpp
void optimalAlpha_cpp(Rcpp::XPtr<Alpha_shape_3> as_xptr, int nc) {
  Alpha_shape_3* as_ptr = as_xptr.get();
  Alpha_iterator opt = (*as_ptr).find_optimal_alpha(nc);
  Rcpp::Rcout << *opt;
}
```

Why I can't dereference the pointer?

**Comment by sloriot** [comment]
You are making a copy. You need to take a reference like
`Alpha_shape_3& as = *as_ptr;`

**Reply by stla** [reply]
Thanks! This code compiles. But I still get the nonsensical optimal alpha (7.35214e+19).

**Comment by stla** [comment]
Here is the full code:

```cpp
Rcpp::XPtr<Alpha_shape_3> AS_cpp(Rcpp::NumericMatrix pts) {
  // make list of points
  const int npoints = pts.ncol();
  std::list<Point3> points;
  for(int i = 0; i < npoints; i++) {
    const Rcpp::NumericVector pt_i = pts(Rcpp::_, i);
    points.push_back(Point3(pt_i(0), pt_i(1), pt_i(2)));
  }
  // compute alpha shape
  Alpha_shape_3 as(points.begin(), points.end());
  // returns R pointer
  Rcpp::XPtr<Alpha_shape_3> as_xptr(&as, false);
  return as_xptr;
}

void optimalAlpha_cpp(Rcpp::XPtr<Alpha_shape_3> as_xptr, int nc) {
  Alpha_shape_3& as = *(as_xptr.get());
  Alpha_iterator opt = as.find_optimal_alpha(nc);
  Rcpp::Rcout << *opt;
}
```

If I apply `AS_cpp` in R, I get the R pointer. Then I pass it to `optimalAlpha_cpp`, and it always prints `7.35214e+19`.

**Reply by sloriot** [reply]
Hard to tell without the point cloud, the alpha and the input parameters.

**Reply by stla** [reply]
The result is always 7.35214e+19, whatever the point cloud (e.g. a cube or any polyhedron). What do you mean by the alpha and the input parameters?
If I compute the optimal alpha inside the `AS_cpp` function, it works.

**Reply by stla** [reply]
If I move the contents of the second function inside the first function, it works.

**Reply by sloriot** [reply]
`as` in the first function is a local variable that gets destroyed when you leave the function. Doing `Alpha_shape_3* as_ptr = new Alpha_shape_3(points.begin(), points.end());` should fix the issue.

**Reply by stla** [reply]
Yes, it works! Thanks!
